### PR TITLE
nsinit: Adds support for exec into docker containers.

### PR DESCRIPTION
**Pull request by mrunalp**
This is a POC for nsinit exec support for docker containers.

@crosbymichael @LK4D4 @vmarmol WDYT about adding this? I can clean it up if we are okay adding this.

Signed-off-by: Mrunal Patel

### CODE REVIEW

**Comment by dqminh** [comment]
I like the feature, but perhaps it can be implemented as a flag to normal `exec` since the code looks almost the same except that we load the state file from docker-specific directory. Something like `nsinit exec --container a560... bash`

**Comment by rjnagal** [comment]
Agree w/ @dqminh. Are there going to be differences from exec other than state file directory ?

**Comment by mrunalp** [comment]
@dqminh @rjnagal Yes, it makes sense to add this as a flag to exec. I will make the changes. Thanks!

**Comment by mrunalp** [comment]
Updated.

File: nsinit/docker.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	return libcontainer.New(dockerRoot, cgm)
+}
+
+func loadState(root string) (*libcontainer.State, error) {
+	f, err := os.Open(filepath.Join(root, stateFilename))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, err
+		}
+		return nil, err
```
Comment (dqminh) [inline nsinit/docker.go]: both returns here is the same. We can probably just have one `return nil, err` instead.

**Comment by dqminh** [comment]
LGTM other than one small nit.

File: nsinit/docker.go (hunk 1 above)
Comment (mrunalp) [inline nsinit/docker.go]: Fixed.

**Comment by crosbymichael** [comment]
@mrunalp what is the usecase for this?

**Comment by mrunalp** [comment]
A light weight way to exec into docker containers and also a debug tool.

**Comment by rjnagal** [comment]
@crosbymichael WDYT about this one? It does make running exec simpler in some cases.

**Comment by crosbymichael** [comment]
idk, i would rather not add docker specific things too this tool

**Comment by mrunalp** [comment]
I will try to bat one more time :) I think there are requirements cropping up for such a utility. It is similar to how we used to be able to enter into a docker container using its container.json.

**Comment by mrunalp** [comment]
We could add a disclaimer that this is not a supported tool and use at your own risk.

**Comment by crosbymichael** [comment]
Why would you not use docker exec or this?

**Comment by ncdc** [comment]
@crosbymichael we're looking for a way to avoid proxying in/out/err through the docker daemon - trying to minimize hops and potential bottlenecks

**Comment by dqminh** [comment]
I like this too for local exec. One can use `nsinit -docker` to bypass the needs of docker daemon and gives the user more control over the exec sessions ( i.e., wait / stop / kill exec sessions properly without resolving to patch the docker API which we do now ). The user admittedly have to take care of shipping a compatible nsinit version with current docker, but that can be taken care of easily.

**Comment by LK4D4** [comment]
@crosbymichael What your thoughts now?

**Pull request closed without merging**